sdk/metric/reader.go

```diff
@@ -0,0 +1,92 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package metric
+
+import (
+	"context"
+
+	"go.opentelemetry.io/otel/sdk/metric/export"
+	"go.opentelemetry.io/otel/sdk/metric/view"
+)
+
+// Reader is the interface used between the SDK and an
+// exporter.  Control flow is bi-directional through the
+// Reader, since the SDK initiates ForceFlush and Shutdown
+// while the initiates collection.  The Register() method here
+// informs the Reader that it can begin reading, signaling the
+// start of bi-directional control flow.
+//
+// Typically, push-based exporters that are periodic will
+// implement PeroidicExporter themselves and construct a
+// PeriodicReader to satisfy this interface.
+//
+// Pull-based exporters will typically implement Register
+// themselves, since they read on demand.
+type Reader interface {
+	// Register is called when the SDK is fully
+	// configured.  The Producer passed allows the
+	// Reader to begin collecting metrics using its
+	// Produce() method.
+	register(producer)
+
+	Collect(context.Context) (export.Metrics, error)
+
+	// ForceFlush is called when MeterProvider.ForceFlush() is called.
+	ForceFlush(context.Context) error
+
+	// Shutdown is called when MeterProvider.Shutdown() is called.
+	Shutdown(context.Context) error
+}
+
+// Producer is the interface used to perform collection by the reader.
+type producer interface {
+	// Produce returns metrics from a single collection.
+	//
+	// Produce may be called concurrently,
+	//
+	// The `in` parameter supports re-use of memory from
+	// one collection to the next.  Callers that pass `in`
+	// will write metrics into the same slices and structs.
+	//
+	// When `in` is nil, a new Metrics object is returned.
+	produce(context.Context) export.Metrics
```

Comment:
Should return an error? Misconfigured instruments, invalid aggregations, and other system failure should likely travel back up the chain. That way a `Reader` might try and handle the error.

Reply:
FYI: the way code is organized in my branch, there are no error conditions that can happen during produce().  All error messages would be put together when the instrument is compiled, and all errors thus return to the user who allocated the instrument (referring to the view and reader name associated with the conflict). Since Produce() is a pure data manipulation, there are no system failures either.

Reply:
What does it look like for a reader to handle this error?  Would that just have us have `otel.Handle(err)` in every reader when we could have it as part of our producer?

Reply:
> What does it look like for a reader to handle this error? Would that just have us have `otel.Handle(err)` in every reader when we could have it as part of our producer?

Having the Reader return it via the `Collect` method and let the user decide sounds like the best approach.

Reply:
After discussing this there can be an error in calling callbacks, eg. timeouts.  So I will add an error to this interface.